src/python/pants/binaries/binary_util.py

```diff
@@ -129,48 +130,67 @@ def __init__(self, baseurls, timeout_secs, bootstrapdir, path_by_id=None):
     if path_by_id:
       self._path_by_id.update((tuple(k), tuple(v)) for k, v in path_by_id.items())
 
-  def select(self, supportdir, version, name, platform_dependent):
-    if platform_dependent:
-      return self._select_binary(supportdir, version, name)
-    else:
-      return self._select_script(supportdir, version, name)
-
-  def select_binary(self, supportdir, version, name):
-    return self._select_binary(supportdir, version, name)
-
-  def select_script(self, supportdir, version, name):
-    return self._select_script(supportdir, version, name)
-
   # TODO: Deprecate passing in an explicit supportdir? Seems like we should be able to
   # organize our binary hosting so that it's not needed.
-  def _select_binary(self, supportdir, version, name):
-    """Selects a binary matching the current os and architecture.
+  def select(self, supportdir, version, name, platform_dependent, archive_type):
+    """Fetches a file, unpacking it if necessary."""
```

Comment:
he usage based on the codes currently will be slightly different for archive 
  * name of the script without extension
  * will only support one extension per archive type, say only .tgz, no .tar.gz
and non-archive
  * full name of the script / file

Reply:
Oops, didn't see this comment before. I mentioned above that it kinda bothers me how `select()` currently returns a path to either a binary or an unpacked archive, if the BinaryTool subclass specifies `archive_type`. It already handles constructing the appropriate filename to fetch (e.g. appending `.tar.gz` when `archive_type='tgz'`) and unpacks the archive before returning the path to the unpacked archive. Right now I think handling the inconsistency in returning a path or a directory will require a small amount of extra code, and is something that should be handled in a subsequent PR which uses that feature to evaluate how simple / appropriate the solution would be (which I'll put up after this gets merged).

Reply:
A task using the `LLVM` subsystem right now would have to know that `select()` is going to return a directory, but that would be clear enough, I think, because we have to specify `archive_type` in the `LLVM` class itself anyway.

Reply:
I think it's OK as-is. We're assuming a naming convention for the paths we fetch from, but that's fine.